experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/YarnApplicationMaster.scala

```diff
@@ -0,0 +1,338 @@
+import org.apache.gearpump.experiments.yarn.CmdLineVars.{APPMASTER_IP, APPMASTER_PORT}
+import org.apache.gearpump.experiments.yarn.Constants._
+import org.apache.gearpump.experiments.yarn._
+import org.apache.gearpump.transport.HostPort
+import org.apache.gearpump.util.LogUtil
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.net.NetUtils
+import org.apache.hadoop.yarn.api.protocolrecords.RegisterApplicationMasterResponse
+import org.apache.hadoop.yarn.api.records.{Container, ContainerId, FinalApplicationStatus}
+import org.apache.hadoop.yarn.client.api.AMRMClient.ContainerRequest
+import org.apache.hadoop.yarn.client.api.async.impl.NMClientAsyncImpl
+import org.apache.hadoop.yarn.client.api.async.{AMRMClientAsync, NMClientAsync}
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+import org.slf4j.Logger
+
+import scala.util.{Failure, Success, Try}
+
+object AmStates {
+
+  sealed trait Reason
+  case class Failed(throwable: Throwable) extends Reason
+  case object ShutdownRequest extends Reason
+  case object AllRequestedContainersCompleted extends Reason
+
+}
+
+object AmActorProtocol {
+  import AmStates._
+
+  case class AMRMClientAsyncStartup(status: Try[Boolean])
+  case class LaunchContainers(containers: List[Container])
+  case class LaunchWorkerContainers(containers: List[Container])
+  case class LaunchServiceContainer(containers: List[Container])
+  case class ContainerRequestMessage(memory: Int, vCores: Int)
+  case class RMHandlerDone(reason: Reason, rMHandlerContainerStats: RMHandlerContainerStats)
+  case class RMHandlerContainerStats(allocated: Int, completed: Int, failed: Int)
+  case class RegisterAMMessage(appHostName: String, appHostPort: Int, appTrackingUrl: String)
+  case class RegisterAppMasterResponse(response: RegisterApplicationMasterResponse)
+  case class AMStatusMessage(appStatus: FinalApplicationStatus, appMessage: String, appTrackingUrl: String)
+  case class ContainerStarted(containerId: ContainerId)
+
+}
+
+
+class YarnApplicationMaster(appConfig: AppConfig, yarnConf: YarnConfiguration,
+                            propsRMClient: Props,
+                            createRMCallbackHandler: (AppConfig, ActorRef) => ResourceManagerCallbackHandler,
+                            createNMClient: (NodeManagerCallbackHandler, YarnConfiguration) => NMClientAsync,
+                            createNodeManagerCallbackHandler: (ActorRef) => NodeManagerCallbackHandler,
+                            containerLaunchContextFactory: (String) => org.apache.hadoop.yarn.api.records.ContainerLaunchContext) extends Actor {
+
+  import AmActorProtocol._
+  import AmStates._
+
+  val LOG: Logger = LogUtil.getLogger(getClass)
+  val nodeManagerCallbackHandler = createNodeManagerCallbackHandler(self)
+  val nodeManagerClient: NMClientAsync = createNMClient(nodeManagerCallbackHandler, yarnConf)
+  val rmClientActor = context.actorOf(propsRMClient, "rmClient")
```

Comment:
Why creating rmClientActor  inside YarnApplicationMaster? Can you pass the rmClientActor  into YarnApplicaitonMaster?